watertap/unit_models/boron_removal.py

```diff
@@ -0,0 +1,1123 @@
+    **MomentumBalanceType.pressurePhase** - pressure balances for each phase,
+    **MomentumBalanceType.momentumTotal** - single momentum balance for material,
+    **MomentumBalanceType.momentumPhase** - momentum balances for each phase.}""",
+        ),
+    )
+
+    CONFIG.declare(
+        "property_package",
+        ConfigValue(
+            default=useDefault,
+            domain=is_physical_parameter_block,
+            description="Property package to use for control volume",
+            doc="""Property parameter object used to define property calculations,
+    **default** - useDefault.
+    **Valid values:** {
+    **useDefault** - use default package from parent model or flowsheet,
+    **PhysicalParameterObject** - a PhysicalParameterBlock object.}""",
+        ),
+    )
+
+    CONFIG.declare(
+        "property_package_args",
+        ConfigBlock(
+            implicit=True,
+            description="Arguments to use for constructing property packages",
+            doc="""A ConfigBlock with arguments to be passed to a property block(s)
+    and used when constructing these,
+    **default** - None.
+    **Valid values:** {
+    see property package for documentation.}""",
+        ),
+    )
+
+    CONFIG.declare(
+        "chemical_mapping_data",
+        ConfigValue(
+            default={},
+            domain=dict,
+            description="Dictionary of chemical species names and their mapping",
+            doc="""
+        Dictionary of chemical species names from the property
+        package and how they map to specific species needed for solving a simple
+        boron speciation problem in an equilibrium reactor. This dictionary must
+        have the following format [Required]::
+
+            {'boron_name': 'name_of_species_representing_boron', #[is required]
+            'borate_name': 'name_of_species_representing_borate', #[is required]
+            'proton_name': 'name_of_species_representing_protons',  #[is optional]
+            'hydroxide_name': 'name_of_species_representing_hydroxides', #[is optional]
+            'caustic_additive':
+                {
+                    'additive_name': 'name_of_the_actual_chemical', #[is optional]
+                    'cation_name': 'name_of_cation_species_in_additive', #[is optional]
+                    'mw_additive': (value, units), #[is required]
+                    'charge_additive': value, #[is required]
+                },
+            }
+
+            """,
+        ),
+    )
+
+    def build(self):
+        # build always starts by calling super().build()
+        # This triggers a lot of boilerplate in the background for you
+        super().build()
+
+        # this creates blank scaling factors, which are populated later
+        self.scaling_factor = Suffix(direction=Suffix.EXPORT)
+
+        # Next, get the base units of measurement from the property definition
+        units_meta = self.config.property_package.get_metadata().get_derived_units
+

```

Comment:
Can this units_meta be removed since all units are/can be adopted from `pyunits` directly?

Reply:
This functionality is intended to allow for general models where the units of measurement are not know a priori - i.e. in many cases the units of measurement are determined by the property package, and thus the units definition in the unit model will need to change depending on what the property package is using.